fabric-events-interaction-v0/src/main/java/net/fabricmc/fabric/api/event/player/BlockBreakEvents.java

```diff
@@ -0,0 +1,56 @@
+package net.fabricmc.fabric.api.event.player;
+
+import net.fabricmc.fabric.api.event.Event;
+import net.fabricmc.fabric.api.event.EventFactory;
+import net.minecraft.block.Block;
+import net.minecraft.block.BlockState;
+import net.minecraft.block.entity.BlockEntity;
+import net.minecraft.entity.player.PlayerEntity;
+import net.minecraft.util.ActionResult;
+import net.minecraft.util.math.BlockPos;
+
+public class BlockBreakEvents {
+	/**
+	 * Callback before a block is broken.
+	 * Only called on the server, however updates are synced with the client.
+	 *
+	 * <p>Upon return:
+	 * <ul><li>SUCCESS/PASS/CONSUME continues the default code for breaking the block
+	 * <li>FAIL cancels the block breaking action</ul>
+	 */
+	public static final Event<BeforeBreakBlockCallback> BEFORE = EventFactory.createArrayBacked(BeforeBreakBlockCallback.class,
+			(listeners) -> (player, pos, state, entity, block) -> {
+				for (BeforeBreakBlockCallback event : listeners) {
+					ActionResult result = event.beforeBlockBreak(player, pos, state, entity, block);
+
+					if (result != ActionResult.FAIL) {
+						return result;
+					}
+				}
+
+				return ActionResult.FAIL;
+			}
+	);
+
+	/**
+	 * Callback after a block is broken.
+	 * Called on both Client and Server
+	 */
+	public static final Event<AfterBreakBlockCallback> AFTER = EventFactory.createArrayBacked(AfterBreakBlockCallback.class,
+			(listeners) -> (player, pos, state, entity, block) -> {
+				for (AfterBreakBlockCallback event : listeners) {
+					event.afterBlockBreak(player, pos, state, entity, block);
+				}
+			}
+	);
+
+	@FunctionalInterface
+	public interface BeforeBreakBlockCallback {
+		ActionResult beforeBlockBreak(PlayerEntity player, BlockPos pos, BlockState state, /* Nullable */ BlockEntity entity, Block block);
```

Comment:
I think we should pass the world in the parameters.

The BlockEntity can be obtained using `getBlockEntity` in world. Also I don't think the block param at the end is really necessary since you can get that from the block state.

Also please document the fields, I have no idea whether the block state is the state being broken or the resulting state.

Reply:
I added world to the parameters, and I am documenting it now. The blockentity is there as the method where the event is injected to already has a blockentity there, so there is no point of the double computation.